src/site/shared/containers/ContextMenu/index.tsx

```diff
@@ -55,7 +58,47 @@ export const ContextMenu = ({info, paste}: Props) => {
             else if (ev.type === "mousedown")
                 dispatch(CloseContextMenu());
         });
-    }, [input])
+    }, [input]);
+
+    /*Position changes are calculated using the react hook so that the
+     *context menu does not jump around during other update events.
+    */
+    useEffect(() => {
+        if(isOpen){
+            let pos = input?.getMousePos();
+
+            if (!menu.current)
+                throw new Error("ContextMenu failed: menu.current is null");
+
+            const offset = 1;
+            const contextMenuWidth = menu.current.getBoundingClientRect().width;
+            const contextMenuHeight = menu.current.getBoundingClientRect().height;
+
+
+            if (pos.x+ contextMenuWidth > window.innerWidth)
+                pos.x -= contextMenuWidth - offset;
+
+            if (pos.y + contextMenuHeight + HEADER_HEIGHT - CONTEXT_MENU_VERT_OFFSET > window.innerHeight)
+                pos.y -= contextMenuHeight - offset;
+
+            //updates position state
+            setPos({ posX:pos.x, posY:pos.y });
+
+            // Update context menu position on canvas
+            menuPos = camera.getWorldPos(input.getMousePos());
```

Comment:
I had to re-remember why this was even necessary, I think it's because the position to `paste` at isn't necessarily the context menu position since if the user clicks near the edge of the screen, the context menu gets moved slightly away.

But anyways, unfortunately, this should also be converted to a state variable and updated accordingly, but let's rename it like `pastePos` (`pastePosX`, `pastePosY`)